pkg/provisioner/ironic/ironic.go

```diff
@@ -715,6 +718,152 @@ func (p *ironicProvisioner) getUpdateOptsForNode(ironicNode *nodes.Node) (update
+	}
+	return
+}
+
+// A private method to set target for RAID configuration steps
+func (p *ironicProvisioner) setTargetRAIDCfg(ironicNode *nodes.Node) (err error) {
+	// Build target for RAID configuration steps
+	logicalDisks, err := p.buildTargetRAIDCfg()
+	if len(logicalDisks) == 0 || err != nil {
+		return
+	}
+	// Set target for RAID configuration steps
+	err = nodes.SetRAIDConfig(
+		p.client,
+		ironicNode.UUID,
+		nodes.RAIDConfigOpts{LogicalDisks: logicalDisks},
+	).ExtractErr()
+	if err != nil {
+		p.log.Error(err, "failed in setting the target RAID configuration",
+			"logical disks", logicalDisks,
+		)
+	}
+	return
+}
+
+// A private method to build the clean steps for RAID configuration from BaremetalHost spec
+func (p *ironicProvisioner) buildRAIDCleanSteps() (cleanSteps []nodes.CleanStep) {
+	// To add ‘delete_configuration’ before ‘create_configuration’ to make sure
+	cleanSteps = append(
+		cleanSteps,
+		nodes.CleanStep{
+			Interface: "raid",
+			Step:      "delete_configuration",
+		},
+	)
+	// If not configure raid, only need to clear old configuration
+	if p.host.Spec.RAID == nil || p.host.Spec.RAID.Volumes == nil || len(p.host.Spec.RAID.Volumes) == 0 {
+		return
+	}
+	if p.host.Spec.RAID.IsSoftwareRaid {
+		cleanSteps = append(
+			cleanSteps,
+			nodes.CleanStep{
+				Interface: "deploy",
+				Step:      "erase_devices_metadata",
+			},
+		)
+	}
+	// ‘create_configuration’ doesn’t remove existing disks. It is recommended
+	// that only the desired logical disks exist in the system after manual cleaning.
+	cleanSteps = append(
+		cleanSteps,
+		nodes.CleanStep{
+			Interface: "raid",
+			Step:      "create_configuration",
+		},
+	)
+	return
+}
+
+func (p *ironicProvisioner) startManualCleaning(ironicNode *nodes.Node) (result provisioner.Result, err error) {
+	cleanSteps := make([]nodes.CleanStep, 0)
+
+	if p.host.Spec.RAID != nil && p.host.Spec.RAID.DeleteExisting {
+		// Set raid config
+		err = p.setTargetRAIDCfg(ironicNode)
+		if err != nil {
+			return
+		}
+		// Build raid configuration clean steps
+		cleanSteps = append(cleanSteps, p.buildRAIDCleanSteps()...)
+	}
+
+	// Start manual clean
+	if len(cleanSteps) != 0 {
+		p.log.Info("remove existing configuration and set new configuration", "steps", cleanSteps)
+		opts := nodes.ProvisionStateOpts{Target: nodes.TargetClean, CleanSteps: cleanSteps}
+		return p.changeNodeProvisionState(ironicNode, opts, false)
+	}
+	return
+}
+
+// Prepare remove existing configuration and set new configuration
+func (p *ironicProvisioner) Prepare() (result provisioner.Result, err error) {
+	var ironicNode *nodes.Node
+
+	if ironicNode, err = p.findExistingHost(); err != nil {
+		return result, errors.Wrap(err, "could not find host to clean")
+	}
+	if ironicNode == nil {
+		return result, fmt.Errorf("no ironic node for host")
+	}
+
+	switch nodes.ProvisionState(ironicNode.ProvisionState) {
+	case nodes.Manageable:
+		return p.startManualCleaning(ironicNode)
```

Comment:
At the least shouldn't we return without setting the Dirty flag if the node is in the Cleaning state? We may fail to write the state to the CR after initially starting the cleaning, so we should do the right thing if we find ourselves back here.

It would be nice to be able to rely on this rather than hack in an extra parameter to `changeNodeProvisionState()`, but I guess that would introduce a timing dependency.

Reply:
There are three cases in the `managerable` state:
- (`Ready` -> `Preparing`) Before `manual cleaning` -> do `manual cleaning`
- (`Preparing` -> `Preparing`) Start `manual cleaning` failed (409) -> do `manual cleaning` again
- (`Preparing` -> `Preparing`) Start `manual clean` success -> go to `provisioning` state
On addition to the current solution, I have not found a solution that is compatible with the three cases. If we can solve it, we also can check `cleaning fail` in `Preparing` state.

Reply:
I replaced `hack in an extra parameter to changeNodeProvisionState ()` by `checking ironicNode.CleanStep`. PTAL.

Reply:
I replaced "hack in an extra parameter to changeNodeProvisionState ()" by "checking ironicNode.CleanStep". PTAL.

Reply:
You're assuming that "-> go to `provisioning` state" can't fail, but it can.